src/data/dhis2/TrackedEntityInstances.ts

```diff
@@ -0,0 +1,217 @@
+import { Ref } from "../../domain/entities/Ref";
+import { D2Api, Id } from "../../types/d2-api";
+import { Moment } from "moment";
+import {
+    RelationshipMetadata,
+    RelationshipOrgUnitFilter,
+    fromApiRelationships,
+    getRelationshipMetadata,
+} from "../repositories/download-template/Dhis2RelationshipTypes";
+import {
+    AttributeValue,
+    Enrollment,
+    Program,
+    TrackedEntityInstance,
+} from "../../domain/entities/TrackedEntityInstance";
+import {
+    PaginatedTeiGetResponse,
+    TrackedEntityInstance as TrackedEntityInstanceApi,
+} from "@eyeseetea/d2-api/api/trackedEntityInstances";
+import { KeysOfUnion } from "../../types/utils";
+import { DataElementType } from "../../domain/entities/DataForm";
+import { Geometry } from "../../domain/entities/Geometry";
+
+export interface GetOptions {
+    api: D2Api;
+    program: Ref;
+    orgUnits: Ref[];
+    pageSize?: number;
+    enrollmentStartDate?: Moment;
+    enrollmentEndDate?: Moment;
+    relationshipsOuFilter?: RelationshipOrgUnitFilter;
+}
+
+export async function getTrackedEntityInstances(options: GetOptions): Promise<TrackedEntityInstance[]> {
+    const {
+        api,
+        orgUnits,
+        pageSize = 500,
+        enrollmentStartDate,
+        enrollmentEndDate,
+        relationshipsOuFilter = "CAPTURE",
+    } = options;
+    if (_.isEmpty(orgUnits)) return [];
+
+    const program = await getProgram(api, options.program.id);
+    if (!program) return [];
+
+    const metadata = await getRelationshipMetadata(program, api, {
+        organisationUnits: orgUnits,
+        ouMode: relationshipsOuFilter,
+    });
+
+    // Avoid 414-uri-too-large by spliting orgUnit in chunks
+    const orgUnitsList = _.chunk(orgUnits, 250);
+
+    // Get TEIs for the first page:
+    const apiTeis: TrackedEntityInstanceApi[] = [];
+
+    for (const orgUnits of orgUnitsList) {
+        // Limit response size by requesting paginated TEIs
+        for (let page = 1; ; page++) {
+            const { pager, trackedEntityInstances } = await getTeisFromApi({
+                api,
+                program,
+                orgUnits,
+                page,
+                pageSize,
+                enrollmentStartDate,
+                enrollmentEndDate,
+                ouMode: relationshipsOuFilter,
+            });
+            apiTeis.push(...trackedEntityInstances);
+            if (pager.pageCount <= page) break;
+        }
+    }
+
+    return apiTeis.map(tei => buildTei(metadata, program, tei));
+}
+
+export async function getProgram(api: D2Api, id: Id): Promise<Program | undefined> {
```

Comment:
already exists in InstanceDefaultRepository: getProgram, please reuse

Reply:
So here I have a question about this one, since this file is not a useCase, would it make sense to make use of a Repository's function? I guess the functionality is the same although written differently, but I kept it as is since I wasn't sure about this one.

Reply:
this function is called from  DownloadTemplateDefaultRepository - we are already calling a repo from a repo. So shouldn't be an issue.